llama.cpp

```diff
@@ -11005,6 +11174,223 @@ struct llm_build_context {
+
+                // split into {n_head * qk_nope_head_dim, n_tokens}
+                struct ggml_tensor * k_nope = ggml_view_3d(ctx0, kv, qk_nope_head_dim, n_head, n_tokens, ggml_element_size(kv) * (qk_nope_head_dim + hparams.n_embd_head_v), ggml_element_size(kv) * n_head * (qk_nope_head_dim + hparams.n_embd_head_v), 0);
+                cb(k_nope, "k_nope", il);
+
+                // and {n_head * n_embd_head_v, n_tokens}
+                struct ggml_tensor * value_states = ggml_view_3d(ctx0, kv, hparams.n_embd_head_v, n_head, n_tokens, ggml_element_size(kv) * (qk_nope_head_dim + hparams.n_embd_head_v), ggml_element_size(kv) * n_head * (qk_nope_head_dim + hparams.n_embd_head_v), ggml_element_size(kv) * qk_nope_head_dim);
+                cb(value_states, "value_states", il);
+
+                value_states = ggml_cont(ctx0, value_states);
+                cb(value_states, "value_states", il);
+
+                value_states = ggml_view_2d(ctx0, value_states, hparams.n_embd_head_v * n_head, n_tokens, ggml_element_size(kv) * hparams.n_embd_head_v * n_head, 0);
+                cb(value_states, "value_states", il);
+
+                q_pe = ggml_rope_ext(
+                    ctx0, q_pe, inp_pos, nullptr,
+                    n_rot, rope_type, 0, n_orig_ctx, freq_base, freq_scale,
+                    ext_factor, attn_factor_scaled, beta_fast, beta_slow
+                );
+                cb(q_pe, "q_pe", il);
+
+                // shared RoPE key
+                k_pe = ggml_rope_ext(
+                    ctx0, ggml_view_3d(ctx0, k_pe, qk_rope_head_dim, 1, n_tokens, k_pe->nb[0], k_pe->nb[1], 0), inp_pos, nullptr,
+                    n_rot, rope_type, 0, n_orig_ctx, freq_base, freq_scale,
+                    ext_factor, attn_factor_scaled, beta_fast, beta_slow
+                );
+                cb(k_pe, "k_pe", il);
+
+                struct ggml_tensor * query_states = ggml_new_tensor_3d(ctx0, q_nope->type, hparams.n_embd_head_k, n_head, n_tokens);
+                cb(query_states, "query_states", il);
+                query_states = ggml_set_inplace(ctx0, query_states, q_nope, query_states->nb[1], query_states->nb[2], query_states->nb[3], 0);
+                query_states = ggml_set_inplace(ctx0, query_states, q_pe, query_states->nb[1], query_states->nb[2], query_states->nb[3], ggml_element_size(query_states) * qk_nope_head_dim);
+
+                k_pe = ggml_repeat(ctx0, k_pe, q_pe);
+                cb(k_pe, "k_pe", il);
+
+                struct ggml_tensor * key_states = ggml_new_tensor_3d(ctx0, q_nope->type, hparams.n_embd_head_k, n_head, n_tokens);
+                cb(key_states, "key_states", il);
+                key_states = ggml_set_inplace(ctx0, key_states, k_nope, key_states->nb[1], key_states->nb[2], key_states->nb[3], 0);
+                key_states = ggml_set_inplace(ctx0, key_states, k_pe, key_states->nb[1], key_states->nb[2], key_states->nb[3], ggml_element_size(key_states) * qk_nope_head_dim);
```

Comment:
And `GGML_OP_SET` as well. I am not quite sure what this op does. Could it be replaced with a `ggml_cpy` into viwes?

Reply:
It should be possible, though I think it would require `ggml_build_forward_expands` which is not very nice

I think this OP does something like this:

```
k_pe:   kkk
q_pe:   qqq qqq qqq
k_nope: nnn

key_states: nnn kkk kkk kkk
```

Reply:
Having `ggml_new_tensor_3d` is not great either, they will be treated as inputs in ggml-alloc and allocated at the beginning of the compute buffer, which will increase memory usage since there is one for each layer.

Reply:
It might be better to spend the time to make a more general `ggml_concat` operations.

Reply:
> It might be better to spend the time to make a more general `ggml_concat` operations.

@ggerganov I agree with @slaren, this could be replaced with a single call to `ggml_concat`, but it would have to allow selecting concatenated dimension (the current implementation concatenates `a->ne[2] + b->ne[2]`, I would need `a->ne[0] + b->ne[0]`. I guess alternatively I could do `ggml_permute` + `ggml_concat` + `ggml_permute`, but it would be equally ugly to the current `ggml_new_tensor_3d` + `ggml_set_inplace` + `ggml_set_inplace`

Reply:
Ok, will try to generalize `ggml_concat` now to match https://pytorch.org/docs/stable/generated/torch.cat.html#torch.cat

Reply:
@ggerganov Does `ggml_concat` support broadcasting? If yes, perhaps we could get rid of the ggml_repeat() call. It's only purpose is to repeat the shared rope key for all attention heads.

Reply:
Hm, how would broadcasting work with `ggml_concat`? This is the current API:

```c
    // concat a and b along dim
    // used in stable-diffusion
    GGML_API struct ggml_tensor * ggml_concat(
            struct ggml_context * ctx,
            struct ggml_tensor  * a,
            struct ggml_tensor  * b,
            int                   dim);
```

Reply:
> Hm, how would broadcasting work with `ggml_concat`? This is the current API:

@ggerganov forget it, it seems that even in numpy and pytorch concat doesn't support broadcasting

Reply:
> If you rebase on `master` and apply the following patch, it should work with CUDA and Metal:
> 
> ```diff
> diff --git a/llama.cpp b/llama.cpp
> index cef5bfdd..9c80a621 100644
> --- a/llama.cpp
> +++ b/llama.cpp
> @@ -11305,18 +11305,11 @@ struct llm_build_context {
>                  );
>                  cb(k_pe, "k_pe", il);
>  
> -                struct ggml_tensor * q_states = ggml_new_tensor_3d(ctx0, q_nope->type, hparams.n_embd_head_k, n_head, n_tokens);
> +                struct ggml_tensor * q_states = ggml_concat(ctx0, q_nope, q_pe, 0);
>                  cb(q_states, "q_states", il);
> -                q_states = ggml_set_inplace(ctx0, q_states, q_nope, q_states->nb[1], q_states->nb[2], q_states->nb[3], 0);
> -                q_states = ggml_set_inplace(ctx0, q_states, q_pe, q_states->nb[1], q_states->nb[2], q_states->nb[3], ggml_element_size(q_states) * n_embd_head_qk_nope);
>  
> -                k_pe = ggml_repeat(ctx0, k_pe, q_pe);
> -                cb(k_pe, "k_pe", il);
> -
> -                struct ggml_tensor * k_states = ggml_new_tensor_3d(ctx0, q_nope->type, hparams.n_embd_head_k, n_head, n_tokens);
> +                struct ggml_tensor * k_states = ggml_concat(ctx0, k_nope, ggml_repeat(ctx0, k_pe, q_pe), 0);
>                  cb(k_states, "k_states", il);
> -                k_states = ggml_set_inplace(ctx0, k_states, k_nope, k_states->nb[1], k_states->nb[2], k_states->nb[3], 0);
> -                k_states = ggml_set_inplace(ctx0, k_states, k_pe, k_states->nb[1], k_states->nb[2], k_states->nb[3], ggml_element_size(k_states) * n_embd_head_qk_nope);
>  
>                  cur = llm_build_kv(ctx0, model, hparams, cparams, kv_self, gf,
>                          model.layers[il].wo, NULL,
> ```

@ggerganov Done, looks much cleaner now, thank you! Works without issues (tested on CPU).

Reply:
Nice, I also tested the Lite model with Metal. Let's merge after @slaren approval